### ISSUE TITLE
Compiler Macro pass

### ISSUE THREAD

**Issue by mhevery** [angular/angular#521]
Have a new `Macro` compiler. This is needed for the i18n work

```
@Macro({
  selector: 'my-button'
})
class MyButton {
  constructor() { // Not injectable.
    // created only once
  }

  transform(el: Element) {
    // called multiple times as the matches happen.
    // can perform DOM transformation on the template
  }
}
```

This will be used in:
- transforming i18n
- transforming `{{exp}}` to `<div [text.0]="exp">` (This should be migrated)

**Comment by tbosch** [comment]
I am not sure whether we should allow Macro directives to use the app
injector. Otherwise we need to run the app on the server, when we
precompile templates on the server. This would require all services to run
on the server as well.

**Comment by mhevery** [comment]
I agree, but not sure how useful Macro directives would be without injector. I guess we can try without injection and than add it later.

How is Macro directive different from the Compiler Step?

**Comment by tbosch** [comment]
In my opinion it is a Compiler Step, although with a nicer API...

**Comment by mhevery** [comment]
Let's punt on this for now and revisit after we have more experience with i18n

**Comment by ajoslin** [comment]
Is there any ETA on this being implemented?

**Comment by gdi2290** [comment]
this would be great for isomorphic since you can create a <form> macro that will decorate/transform it to provide the correct legacy `action` attributes etc while keeping it transparent to the user. the other benefit is that you wouldn't need to include the isomorphic macros on the client since they're only needed for the server. The same goes for SEO macros where you would include the correct itemprop from schema https://schema.org/docs/gs.html while also being transparent to the user

**Comment by mhevery** [comment]
This is obsolete, and no longer relevant / actionable. To keep our issues clean, we are aggressively closing them.
